### PR TITLE
Add Blazor support to Consolonia.

### PR DESCRIPTION
**Pull request by tomlm**
# TLDR
* Use [⚡ Blazor](https://dotnet.microsoft.com/en-us/apps/aspnet/web-apps/blazor) syntax for [Avalonia](https://avaloniaui.net/) apps
* 😎 Simpler syntax than XAML
* 🪄 IntelliSense support
* Get free [🔥 Hot Reload](https://devblogs.microsoft.com/dotnet/introducing-net-hot-reload/) support on-top
* See https://github.com/Epictek/Avalonia-Blazor-Bindings

# Details
* Added **Consolonia.Blazor** library which makes it so you can call **AppBuilder.UseConsoloniaBlazor() **
* Added **Example.Blazor** project which is consolonia app hosting of blazor components 
* Added **Example.Blazor.Components** project which implements MainPage.razor and SubPage.razor

## Example Razor file
Things to note:
* Method binding to code behind
* Conditional layout of elements
* Full intellisense for renaming going to code behind autocomplete etc.
* injection of **INavigation** so you can push and pop views off of the nav stack
* injection of **IClassicDesktopStyleApplicationLifetime** so you can access Args and shutdown the app.
* HOT RELOAD. simple save the file and your view gets rebuilt and rerendered!
 
blazor

```razor
@page "/"
@inject INavigation navigation
@inject IClassicDesktopStyleApplicationLifetime lifetime

@namespace Example.Blazor.Components

<Window Title="Blazor Bindings for Consolonia" Topmost="true">
     <StackPanel Orientation="Orientation.Vertical" VerticalAlignment="VerticalAlignment.Center">
        @if (showCounter)
        {
            <TextBlock HorizontalAlignment="HorizontalAlignment.Center">Counter: @CounterText</TextBlock>
        }
        <StackPanel Orientation="Orientation.Vertical" HorizontalAlignment="HorizontalAlignment.Center" Margin="@Thickness.Parse("4")">
            <Button OnClick="@OnIncrement">Increment counter support</Button>
            <Button OnClick="@OnToggleCounter">@ToggleText</Button>
            <Button OnClick="@OnMessageBox">Message Box</Button>
            <Button OnClick="@OnGotoSubPage">Go to SubPage</Button>
            <Button OnClick="@OnExit">Exit</Button>
        </StackPanel>
    </StackPanel> 
</Window>

@code {
    int counter = 0;
    bool showCounter = true;

    string CounterText => counter switch
    {
        0 => "Not clicked",
        1 => $"Clicked 1 time",
        _ => $"Clicked {counter} times"
    };

    string ToggleText => showCounter ? "Hide Counter" : "Show Counter";

    void OnIncrement()
        => counter++;

    void OnToggleCounter()
        => showCounter = !showCounter;

    void OnGotoSubPage()
        => navigation.PushAsync<SubPage>();

    async void OnMessageBox(RoutedEventArgs args)
    {
        await new MessageBox()
            .ShowDialogAsync((AC.Control)args.Source!, "Hello, Blazor!", "Blazor Bindings for Consolonia");
    }

    void OnExit()
       => lifetime.Shutdown();
}
```

### CODE REVIEW

File: src/Consolonia.Blazor/Consolonia.Blazor.csproj [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+    <ItemGroup>
+        <ProjectReference Include="..\Consolonia.Core\Consolonia.Core.csproj" />
+    </ItemGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Avalonia" Version="$(AvaloniaVersion)" />
+        <PackageReference Include="Avalonia.Desktop" Version="$(AvaloniaVersion)" />
+        <PackageReference Include="BlazorBindingsAvalonia" Version="0.1.2" />
+        <PackageReference Include="Consolonia.Core" Version="11.2" />
```
Comment (goe5mu) [inline src/Consolonia.Blazor/Consolonia.Blazor.csproj:13]: Why PackageReference, not ProjectReference?

File: src/Consolonia.Core/Infrastructure/ConsoloniaNotSupportedException.cs [hunk 2]
```diff
@@ -11,12 +10,6 @@ internal ConsoloniaNotSupportedException(NotSupportedRequest request)
             Request = request;
         }
 
-        protected ConsoloniaNotSupportedException(
```
Comment (goe5mu) [inline src/Consolonia.Core/Infrastructure/ConsoloniaNotSupportedException.cs:14]: Then [Serializable] also can be removed?

File: src/Consolonia.Designer/Consolonia.Designer.csproj [hunk 3]
```diff
@@ -8,6 +8,8 @@
         <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Desktop" Version="$(AvaloniaVersion)" />
         <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Fonts.Inter" Version="$(AvaloniaVersion)" />
         <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
+        <PackageReference Include="Consolonia.Core" Version="11.2" />
```
Comment (goe5mu) [inline src/Consolonia.Designer/Consolonia.Designer.csproj:11]: Why are we referencing as a package if already referenced as a project?

File: src/Consolonia.Designer/Consolonia.Designer.csproj [hunk 4]
```diff
@@ -8,6 +8,8 @@
...
+        <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Consolonia.PreviewHost" Version="11.2" />
```
Comment (goe5mu) [inline src/Consolonia.Designer/Consolonia.Designer.csproj:12]: Why only for DEBUG configuration?

File: src/Consolonia.NUnit/Consolonia.NUnit.csproj [hunk 5]
```diff
@@ -7,6 +7,7 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <PackageReference Include="Consolonia.Core" Version="11.2" />
```
Comment (goe5mu) [inline src/Consolonia.NUnit/Consolonia.NUnit.csproj:10]: + here

File: src/Consolonia.PlatformSupport/Consolonia.PlatformSupport.csproj [hunk 6]
```diff
@@ -1,6 +1,11 @@
 <Project Sdk="Microsoft.NET.Sdk">
     <Import Project="$([MSBuild]::GetPathOfFileAbove('Directory.Core.Build.props', '$(MSBuildThisFileDirectory)../'))" />
-    
+
+    <ItemGroup>
+        <PackageReference Include="Consolonia.Core" Version="11.2" />
+        <PackageReference Include="Consolonia.GuiCS" Version="11.2" />
```
Comment (jinek) [inline src/Consolonia.PlatformSupport/Consolonia.PlatformSupport.csproj:6]: +

File: src/Example.Blazor/Example.Blazor.csproj [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk.Razor">
+
+    <PropertyGroup>
+        <OutputType>Exe</OutputType>
+        <TargetFramework>net8.0</TargetFramework>
```
Comment (jinek) [inline src/Example.Blazor/Example.Blazor.csproj:5]: Should be picked up from Build.props

File: src/Example.Blazor/Example.Blazor.csproj [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Consolonia.Blazor" Version="11.2" />
+        <PackageReference Include="Consolonia.Themes.TurboVision" Version="11.2" />
+        <PackageReference Include="Consolonia.PlatformSupport" Version="11.2" />
```
Comment (jinek) [inline src/Example.Blazor/Example.Blazor.csproj:13]: +

File: src/Example.Blazor.Components/Example.Blazor.Components.csproj [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk.Razor">
+
+    <PropertyGroup>
+        <TargetFramework>net8.0</TargetFramework>
```
Comment (jinek) [inline src/Example.Blazor.Components/Example.Blazor.Components.csproj:4]: + from build.props

File: src/Example.Blazor.Components/Example.Blazor.Components.csproj [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+        <Nullable>enable</Nullable>
+        <ImplicitUsings>enable</ImplicitUsings>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Consolonia.Blazor" Version="11.2" />
```
Comment (jinek) [inline src/Example.Blazor.Components/Example.Blazor.Components.csproj:10]: + why

File: src/Consolonia.Designer/Consolonia.Designer.csproj (hunk 3 above)
Comment (tomlm) [reply src/Consolonia.Designer/Consolonia.Designer.csproj:11]: This is the way you used to do project and package references. Now nuget always treats project references as package references, I have removed this in the packages PR, but I Will do it here too.

File: src/Consolonia.Designer/Consolonia.Designer.csproj (hunk 4 above)
Comment (tomlm) [reply src/Consolonia.Designer/Consolonia.Designer.csproj:12]: also removed in packages pr, but I will do it here too.

File: src/Consolonia.NUnit/Consolonia.NUnit.csproj (hunk 5 above)
Comment (tomlm) [reply src/Consolonia.NUnit/Consolonia.NUnit.csproj:10]: obsolete

File: src/Example.Blazor.Components/Example.Blazor.Components.csproj [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk.Razor">
+
+    <PropertyGroup>
+        <TargetFramework>net8.0</TargetFramework>
```
Comment (jinek) [inline src/Example.Blazor.Components/Example.Blazor.Components.csproj:4]: net version can be removed as it's already mentioned in directory.build.prpos

File: src/Tools/Consolonia.PreviewHost/Properties/launchSettings.json [hunk 12]
```diff
@@ -1,13 +0,0 @@
-{
-  "profiles": {
-    "WSL": {
-      "commandName": "WSL2",
-      "distributionName": ""
-    },
-    "Consolonia.PreviewHost": {
-      "commandName": "Project",
-      "commandLineArgs": ".",
-      "workingDirectory": "S:\\github\\Consolonia\\src\\Consolonia.Gallery"
```
Comment (jinek) [inline src/Tools/Consolonia.PreviewHost/Properties/launchSettings.json:10]: should be perged also? Here is the path hardcoded

**Review by jinek — APPROVED** [review]

**Comment by tomlm** [comment]
sorry I had merge conflicts I had to resolve.

**Review by jinek — APPROVED** [review]
